### DISCUSSION TITLE
The best practice of hierachy with bus interface.

### DISCUSSION THREAD

**Discussion by Readon** [SpinalHDL/SpinalHDL#836] [Q&A]
I have a custom controller which defines two register with RegIf, and it is wrapped into other components, the hierarchy would be as below.
```
top --
     |-component1--
                         |-component2--
                                               |-controller  //here it is, the logic is actually defined here, including how many registers it should use.
```
I defined driveFrom function for the controller to specify the registers it uses with newReg.
However, when I call driveFrom function from top level then it will cause hierarchy violation due to cross level access.

The solution I can imagine is to remain the logic into controller still, but expose the signal need to be changed by RegIf to component1, so the signals can be access from top level. This solution is terrible when the wrapping depth is large.

The problem is caused by the conflict of design rules and convenience.  Is there a better way to balance those two?

**Comment by Dolu1990** [comment]
In which component is defined the driveFrom function ?

I'm currently not sure what are component1 and component2.

Else, using Area instead of Component can help with this.

**Reply by Readon** [reply]
The driveFrom is defined in controller component, because it takes the responsibility for usage of signals.

component1 and component2 are wrapping components which add bus related facilities.

There are many cases to use controller, it might be a top module including component2 directly, or including component1.
Then when to define an Area and when a Component?

**Reply by Dolu1990** [reply]
So i would say, at least, Component for stuff instanciated in the toplevel, then it realy depend the scale of things.

Also, one alternative to component2, maybe to define functions which transform the controller IO into something else. Functions which return area.

In some way, prefering composition over too much component levels.
